openshift/machine-config-operator
# MCO-703: Lifecycle Buildah with MCO

**- What I did**

This PR does the following:

- Consolidates, repackages, and refactors how we parse the data from both the `machine-config-osimageurl` and the `machine-config-operator-images` ConfigMaps for easier code reuse.
- Adds Buildah to the MCO image along with a `build` user.
- Modifies BuildController so that instead of using a hardcoded image pullspec for the builder container, the `image-build` container will now use the same image pullspec as the rest of the MCO components instead. It's source of truth for this value will be the `machine-config-operator-images` ConfigMap.
- Cleaned up some of the constants that were used in BuildController and `pkg/controller/common/constants.go` that are no longer needed, as well as renamed a few of the OCL-specific constants for the annotations and labels used by OCL objects.
- Adds functions to generate label selectors for the aforementioned labels to make querying easier.
- Removes the `TestMachineOSBuilder` e2e test for the following reason: Although the test has been passing for quite some time, ever since the OCL API landed, it has been passing for the incorrect reason. Before the OCL API landed, it was passing because we were looking for a specific ConfigMap to be present and asserting that something would either happen or not happen as a result. In its current form, we assert that nothing should happen. At this point however, we've removed the code that expects a certain ConfigMap to be present. However, this test was never updated to reflect that. With the other changes in this PR, we broke this test by preventing it from compiling. Regardless of that, the test itself does not add much value, so it should be removed.
- Refactors the OCL E2E tests to allow the ImageStream to be created in a different namespace. This enables us to ensure that the correct image pull secret lands on the node as well as ensures that the image pull secret that is bound to the service account is refreshed before each test since we tear down the namespace and its contents upon successful completion of the test.
- Adds labels and annotations to objects created by the OCL E2E test suite to facilitate easier cleanup.

**- How to verify it**

There are a couple of ways to verify this:
1. The MCO image should now contain the `buildah` binary.
2. The `e2e-gcp-op-techpreview` test suite should continue to pass as-is, though I've added some additional assertions.
3. When running the `e2e-gcp-op-techpreview` test suite, the image pullspec for the `image-build` container in the `machine-os-builder` pod should be the same as the rest of the MCO components.
4. All of the objects created by the modified e2e test will have an annotation and label whose key is `machineconfiguration.openshift.io/used-by-e2e-test`. The annotation will have the name of the test currently running.

**- Description for the changelog**
Lifecycle Buildah with MCO
